source_pubsub/client.go

```diff
@@ -255,39 +259,56 @@ func (c *PubSubClient) Wait(ctx context.Context) error {
 	}
 }
 
+func (c *PubSubClient) ResetRetryCount() {
+	c.retryCount = c.maxRetries
+}
+
 func (c *PubSubClient) ReplayID() []byte {
 	return c.currReplayID
 }
 
+func (c *PubSubClient) retryAuth(ctx context.Context, retry bool) (error, bool) {
+	for retry {
+		var err error
+		sdk.Logger(ctx).Info().Msgf("retry connection - retries %d ", c.retryCount)
+
+		if err = c.login(ctx); err != nil && c.retryCount <= 0 {
+			return fmt.Errorf("failed to refresh auth: %w", err), retry
+		} else if err != nil {
+			sdk.Logger(ctx).Info().Msgf("received error on login - retry - %d ", c.retryCount)
+			c.retryCount--
+		}
+
+		if err = c.Initialize(ctx); err != nil && c.retryCount <= 0 {
+			return fmt.Errorf("failed to reinitialize client: %w", err), retry
+		} else if err != nil {
+			sdk.Logger(ctx).Info().Msgf("received error on init - retry - %d ", c.retryCount)
+			c.retryCount--
+		}
+
+		if err == nil {
+			retry = false
+			c.retryCount--
+		}
+	}
+	return nil, retry
+}
+
 func (c *PubSubClient) startCDC(ctx context.Context) error {
 	sdk.Logger(ctx).Info().Msg("starting CDC processing..")
 
 	var (
-		retryAuth   bool
-		retryConn   bool
+		retry       bool
+		err         error
 		lastRecvdAt = time.Now().UTC()
 	)
 
 	for {
-		if retryAuth {
-			sdk.Logger(ctx).Info().Msg("refreshing auth token")
-			if err := c.login(ctx); err != nil {
-				return fmt.Errorf("failed to refresh auth: %w", err)
-			}
-			retryAuth = false
-		}
-
-		if retryConn {
-			sdk.Logger(ctx).Info().Msg("reconnecting to CDC")
-			if err := c.Close(ctx); err != nil {
-				sdk.Logger(ctx).Error().Err(err).Msg("failed to close connection")
-			}
-
-			if err := c.Initialize(ctx); err != nil {
-				return fmt.Errorf("failed to reinitialize client: %w", err)
+		if retry {
```

Comment:
you will need to set retry to false after retry is attempted, otherwise it will retry forever.